src/NumericalAlgorithms/FiniteDifference/Reconstruct.hpp

```diff
@@ -0,0 +1,66 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <utility>  // for std::pair
+
+#include "Utilities/Gsl.hpp"
+
+/// \cond
+template <size_t Dim>
+class Direction;
+template <size_t Dim, typename T>
+class DirectionMap;
+template <size_t Dim>
+class Index;
+/// \endcond
+
+/// \ingroup FiniteDifferenceGroup
+/// Functions and classes for finite difference methods.
+namespace fd {
+/*!
+ * \ingroup FiniteDifferenceGroup
+ * \brief Variable and flux vector splitting reconstruction schemes for finite
+ * difference methods.
+ *
+ * Reconstruction methods that are implemented must use the
+ * `fd::reconstruction::detail::reconstruct` function to perform the
+ * reconstruction. This function performs the actual loops taking into account
+ * strides and ghost zones for the reconstruction method. The type of
+ * reconstruction done is specified with the `Reconstructor` explicit template
+ * parameter. Parameters for the reconstruction scheme are forwarded along using
+ * the `args_for_reconstruction` parameter pack.
+ *
+ * `Reconstructor` classes must define:
+ * - a `static constexpr size_t stencil_width()` function that
+ *   returns the size of the stencil, e.g. 2 for minmod, and 5 for 5-point
+ *   reconstruction.
+ * - a
+ *   \code
+ *      SPECTRE_ALWAYS_INLINE static std::array<double, 2> pointwise(
+ *            const double* const q, const int stride)
+ *   \endcode
+ *   function that optionally takes the additional arguments. The values being
+ *   reconstructed are `q` with \f$q_i\f$ (in the current cell) being `q[0]`.
+ *   The values in neighbor cells, say \f$q_1\f$ is then given by `q[stride]`
+ *   and \f$q_{-1}\f$ is given by `q[-stride]`. The returned values are the
```

Comment:
:+1:

The reason I decided to keep the stride is because I don't know what is the fastest/most efficient way to do reconstruction. I opted for doing the transposes because I failed to get the indexing and strides correct for the eta and zeta reconstruction :P Would it be helpful if I added something like:
```
\note currently the stride is always one because we transpose the data before reconstruction. However,
it may be faster to have a non-unit stride without the transpose. We have the `stride` parameter to make
testing performance easier in the future.
```

Reply:
Yeah, I like this note 👍